### DISCUSSION TITLE
Renovate uses NPM v6 regardless of settings

### DISCUSSION THREAD

**Discussion by JrobT** [renovatebot/renovate#15595] [Request Help] [answered]
Hi!

Thanks for Renovate, it's a great tool. Recently I've rolled out Renovate self-hosted via the `renovate/renovate` Docker image on Bitbucket.

I've been having an issue with some of our Node projects which contain the following in the `package.json` file along with `engine-strict=true`:

```json
"engines": {
  "node": "14",
  "npm": "^8.2.0"
}
```

Renovate will then try to use NPM v6 to update the project's dependencies resulting in failing builds due to the version.

I've had a good look at the issue before coming to you and found a few similar reports/issues, but nothing that I'm 100% sure meets to same situation (maybe you can correct me on this):

-  https://docs.renovatebot.com/self-hosted-configuration/#binarysource & https://docs.renovatebot.com/configuration-options/#constraints
- https://github.com/renovatebot/renovate/discussions/14562
- https://github.com/renovatebot/renovate/issues/12068

I've tried:
- Setting `binarySource` and `constraints` to various values, nothing worked
- Manually installing the latest version of NPM on the `renovate/renovate` docker image before a run with `binarySource=global`, and it still used v6

I thought I'd post as I've run out of ideas with it and wanted to check and see if there was a workaround I'm not thinking of.

**Comment by viceice** [comment, marked as answer]
`binarySource=docker` (needs a docker daemon awailable to renovate) is the only option for `npm` manager to respect contraints yet. work in progress for `binarySource=install`

**Reply by JrobT** [reply]
Thanks for your quick and helpful response! Do you have any example in hand or a link to a doc I can use as a base?

**Reply by viceice** [reply]
depends on which platform (kubernetes / jenkins / droneci / github-actions ...) you are running your renovate

**Reply by JrobT** [reply]
DroneCI. I'm not sure we would want to give daemon access but I would be willing to try it.

**Reply by viceice** [reply]
you can also use docker-in-docker, so you don't need host docker access.

https://docs.drone.io/pipeline/docker/examples/services/docker_dind/

**Reply by viceice** [reply]
remember to use the renovate slim image with docker mode, so you save a lot download size 😉

**Reply by JrobT** [reply]
Great, thank you! I will look at this shortly 👍🏼

**Reply by BenjaminWFox-MuleSoft** [reply]
@viceice Has support for this been released now? I think I'm facing the same issue, and [noticed the docs](https://docs.renovatebot.com/self-hosted-configuration/#binarysource) now note `npm` as supported for Dynamic Install.

**Reply by viceice** [reply]
npm yes, but not node

**Reply by JrobT** [reply]
I ended up creating a base image to upgrade the node/npm versions.
